File: toy-c-compiler-master/src/main/java/toyc/ir/IRBuilder.java
package toyc.ir;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;
import java.util.OptionalInt;
import toyc.common.Type;
import toyc.frontend.ast.ASTNode;
import toyc.frontend.ast.Decl;
import toyc.frontend.ast.Expr;
import toyc.frontend.ast.FuncDef;
import toyc.frontend.ast.Program;
import toyc.frontend.ast.Stmt;
import toyc.ir.block.BasicBlock;
import toyc.ir.block.Function;
import toyc.ir.block.Module;
import toyc.ir.inst.Alloca;
import toyc.ir.inst.BinaryOp;
import toyc.ir.inst.Branch;
import toyc.ir.inst.Call;
import toyc.ir.inst.Compare;
import toyc.ir.inst.CondBranch;
import toyc.ir.inst.GlobalAddr;
import toyc.ir.inst.Instruction;
import toyc.ir.inst.Load;
import toyc.ir.inst.LoadImm;
import toyc.ir.inst.Return;
import toyc.ir.inst.Store;
import toyc.ir.inst.UnaryOp;
import toyc.ir.value.Constant;
import toyc.ir.value.GlobalVar;
import toyc.ir.value.IRValue;
import toyc.ir.value.Label;
import toyc.ir.value.LocalVar;
import toyc.ir.value.Temp;
import toyc.semantic.FunctionSymbol;
import toyc.semantic.SemanticResult;
import toyc.semantic.Symbol;

public final class IRBuilder {
    private final Program program;
    private final SemanticResult semanticResult;
    private final Module module = new Module();
    private final Map<Symbol, IRValue> symbolValues = new IdentityHashMap<>();
    private final Deque<LoopTarget> loopTargets = new ArrayDeque<>();
    private Function currentFunction;
    private BasicBlock currentBlock;
    private int tempCount;
    private int labelCount;
    private int localCount;

    private IRBuilder(Program program, SemanticResult semanticResult) {
        this.program = program;
        this.semanticResult = semanticResult;
    }

    public static IRProgram build(Program program, SemanticResult semanticResult) {
        return new IRProgram(new IRBuilder(program, semanticResult).buildModule());
    }

    public Module buildModule() {
        for (ASTNode def : program.defs()) {
            if (def instanceof Decl decl) {
                emitGlobalDecl(decl);
            }
        }
        for (ASTNode def : program.defs()) {
            if (def instanceof FuncDef funcDef) {
                emitFunction(funcDef);
            }
        }
        return module;
    }

    public Temp newTemp(Type type) {
        return new Temp(tempCount++, type);
    }

    public Label newLabel(String hint) {
        return new Label(hint + "." + labelCount++);
    }

    public LocalVar newLocal(String sourceName, boolean parameter) {
        LocalVar local = new LocalVar(sourceName, localCount++, parameter);
        if (currentFunction != null) {
            currentFunction.addLocal(local);
        }
        return local;
    }

    public BasicBlock appendBlock(String hint) {
        BasicBlock block = new BasicBlock(newLabel(hint));
        currentFunction.addBlock(block);
        return block;
    }

    public void setCurrentBlock(BasicBlock block) {
        currentBlock = block;
    }

    public void emit(Instruction instruction) {
        currentBlock.addInstruction(instruction);
    }

    public void emitBranchIfOpen(Label target) {
        if (!currentBlock.isTerminated()) {
            emit(new Branch(target));
        }
    }

    public IRValue emitExpr(Expr expr) {
        if (expr instanceof Expr.IntLiteral literal) {
            return emitLoadImm(literal.value());
        }
        if (expr instanceof Expr.Id id) {
            return emitId(id);
        }
        if (expr instanceof Expr.Unary unary) {
            return emitUnary(unary);
        }
        if (expr instanceof Expr.Binary binary) {
            return emitBinary(binary);
        }
        if (expr instanceof Expr.FuncCall call) {
            return emitCall(call);
        }
        throw new IllegalArgumentException("unsupported expression: " + expr.getClass().getSimpleName());
    }

    public void emitStmt(Stmt stmt) {
        if (stmt instanceof Decl decl) {
            emitLocalDecl(decl);
        } else if (stmt instanceof Stmt.Block block) {
            emitBlock(block);
        } else if (stmt instanceof Stmt.Empty) {
            return;
        } else if (stmt instanceof Stmt.ExprStmt exprStmt) {
            emitExpr(exprStmt.expr());
        } else if (stmt instanceof Stmt.Assign assign) {
            emitAssign(assign);
        } else if (stmt instanceof Stmt.If ifStmt) {
            emitIf(ifStmt);
        } else if (stmt instanceof Stmt.While whileStmt) {
            emitWhile(whileStmt);
        } else if (stmt instanceof Stmt.Break) {
            emitBreak();
        } else if (stmt instanceof Stmt.Continue) {
            emitContinue();
        } else if (stmt instanceof Stmt.Return ret) {
            emitReturn(ret);
        } else {
            throw new IllegalArgumentException("unsupported statement: " + stmt.getClass().getSimpleName());
        }
    }

    private void emitGlobalDecl(Decl decl) {
        Symbol symbol = semanticResult.symbolOf(decl);
        int initialValue = constValueOrZero(decl instanceof Decl.VarDecl varDecl ? varDecl.init()
                : ((Decl.ConstDecl) decl).init());
        if (decl instanceof Decl.ConstDecl) {
            Constant constant = Constant.named(symbol.name(), initialValue);
            module.addGlobalConstant(constant);
            symbolValues.put(symbol, constant);
        } else {
            GlobalVar global = new GlobalVar(symbol.name(), initialValue);
            module.addGlobal(global);
            symbolValues.put(symbol, global);
        }
    }

    private void emitFunction(FuncDef funcDef) {
        tempCount = 0;
        localCount = 0;
        List<LocalVar> parameters = new ArrayList<>();
        for (FuncDef.Param param : funcDef.params()) {
            LocalVar local = newLocal(param.id(), true);
            parameters.add(local);
            symbolValues.put(semanticResult.symbolOf(param), local);
        }
        currentFunction = new Function(funcDef.id(), toCommonType(funcDef.returnType()), parameters, newLabel(funcDef.id() + ".entry"));
        module.addFunction(currentFunction);
        currentBlock = currentFunction.entryBlock();
        for (LocalVar parameter : parameters) {
            emit(new Alloca(parameter));
        }
        emitBlock(funcDef.body());
        if (!currentBlock.isTerminated()) {
            IRValue defaultValue = currentFunction.returnType() == Type.VOID ? null : emitLoadImm(0);
            emit(new Return(defaultValue));
        }
        currentFunction = null;
        currentBlock = null;
    }

    private void emitBlock(Stmt.Block block) {
        for (Stmt stmt : block.stmts()) {
            if (currentBlock.isTerminated()) {
                break;
            }
            emitStmt(stmt);
        }
    }

    private void emitLocalDecl(Decl decl) {
        Symbol symbol = semanticResult.symbolOf(decl);
        if (decl instanceof Decl.ConstDecl constDecl) {
            int value = constValueOrZero(constDecl.init());
            symbolValues.put(symbol, Constant.named(symbol.name(), value));
            return;
        }
        Decl.VarDecl varDecl = (Decl.VarDecl) decl;
        LocalVar local = newLocal(varDecl.id(), false);
        symbolValues.put(symbol, local);
        emit(new Alloca(local));
        emit(new Store(emitExpr(varDecl.init()), local));
    }

    private void emitAssign(Stmt.Assign assign) {
        IRValue address = addressOf(semanticResult.symbolOf(assign));
        emit(new Store(emitExpr(assign.expr()), address));
    }

    private void emitIf(Stmt.If ifStmt) {
        BasicBlock thenBlock = appendBlock("if.then");
        BasicBlock elseBlock = ifStmt.elseStmt() == null ? null : appendBlock("if.else");
        BasicBlock endBlock = appendBlock("if.end");
        emitCondExpr(ifStmt.cond(), thenBlock.label(), elseBlock == null ? endBlock.label() : elseBlock.label());

        setCurrentBlock(thenBlock);
        emitStmt(ifStmt.thenStmt());
        emitBranchIfOpen(endBlock.label());

        if (elseBlock != null) {
            setCurrentBlock(elseBlock);
            emitStmt(ifStmt.elseStmt());
            emitBranchIfOpen(endBlock.label());
        }

        setCurrentBlock(endBlock);
    }

    private void emitWhile(Stmt.While whileStmt) {
        BasicBlock condBlock = appendBlock("while.cond");
        BasicBlock bodyBlock = appendBlock("while.body");
        BasicBlock endBlock = appendBlock("while.end");
        emit(new Branch(condBlock.label()));

        setCurrentBlock(condBlock);
        emitCondExpr(whileStmt.cond(), bodyBlock.label(), endBlock.label());

        loopTargets.push(new LoopTarget(condBlock.label(), endBlock.label()));
        setCurrentBlock(bodyBlock);
        emitStmt(whileStmt.body());
        emitBranchIfOpen(condBlock.label());
        loopTargets.pop();

        setCurrentBlock(endBlock);
    }

    private void emitBreak() {
        emit(new Branch(loopTargets.peek().breakTarget()));
    }

    private void emitContinue() {
        emit(new Branch(loopTargets.peek().continueTarget()));
    }

    private void emitReturn(Stmt.Return ret) {
        emit(new Return(ret.expr() == null ? null : emitExpr(ret.expr())));
    }

    private IRValue emitId(Expr.Id id) {
        IRValue value = symbolValues.get(semanticResult.symbolOf(id));
        if (value instanceof Constant constant) {
            return emitLoadImm(constant.value());
        }
        Temp result = newTemp(Type.INT);
        emit(new Load(result, addressOf(value)));
        return result;
    }

    private IRValue emitUnary(Expr.Unary unary) {
        IRValue value = emitExpr(unary.expr());
        return switch (unary.op()) {
            case PLUS -> value;
            case MINUS -> {
                Temp result = newTemp(Type.INT);
                emit(new UnaryOp(result, UnaryOp.Op.NEG, value));
                yield result;
            }
            case NOT -> {
                Temp result = newTemp(Type.INT);
                emit(new UnaryOp(result, UnaryOp.Op.NOT, value));
                yield result;
            }
        };
    }

    private IRValue emitBinary(Expr.Binary binary) {
        if (binary.op() == Expr.BinaryOp.AND) {
            return emitShortCircuitAnd(binary);
        }
        if (binary.op() == Expr.BinaryOp.OR) {
            return emitShortCircuitOr(binary);
        }

        IRValue left = emitExpr(binary.left());
        IRValue right = emitExpr(binary.right());
        Temp result = newTemp(Type.INT);
        switch (binary.op()) {
            case ADD -> emit(new BinaryOp(result, BinaryOp.Op.ADD, left, right));
            case SUB -> emit(new BinaryOp(result, BinaryOp.Op.SUB, left, right));
            case MUL -> emit(new BinaryOp(result, BinaryOp.Op.MUL, left, right));
            case DIV -> emit(new BinaryOp(result, BinaryOp.Op.DIV, left, right));
            case MOD -> emit(new BinaryOp(result, BinaryOp.Op.MOD, left, right));
            case LT -> emit(new Compare(result, Compare.Predicate.LT, left, right));
            case GT -> emit(new Compare(result, Compare.Predicate.GT, left, right));
            case LE -> emit(new Compare(result, Compare.Predicate.LE, left, right));
            case GE -> emit(new Compare(result, Compare.Predicate.GE, left, right));
            case EQ -> emit(new Compare(result, Compare.Predicate.EQ, left, right));
            case NEQ -> emit(new Compare(result, Compare.Predicate.NE, left, right));
            case AND, OR -> throw new IllegalStateException("short-circuit operator should be handled earlier");
        }
        return result;
    }

    private IRValue emitShortCircuitAnd(Expr.Binary binary) {
        Temp result = newTemp(Type.INT);
        emit(new LoadImm(result, 0));
        BasicBlock rightBlock = appendBlock("land.rhs");
        BasicBlock endBlock = appendBlock("land.end");
        emit(new CondBranch(emitExpr(binary.left()), rightBlock.label(), endBlock.label()));

        setCurrentBlock(rightBlock);
        emit(new Compare(result, Compare.Predicate.NE, emitExpr(binary.right()), Constant.of(0)));
        emitBranchIfOpen(endBlock.label());

        setCurrentBlock(endBlock);
        return result;
    }

    private IRValue emitShortCircuitOr(Expr.Binary binary) {
        Temp result = newTemp(Type.INT);
        emit(new LoadImm(result, 1));
        BasicBlock rightBlock = appendBlock("lor.rhs");
        BasicBlock endBlock = appendBlock("lor.end");
        emit(new CondBranch(emitExpr(binary.left()), endBlock.label(), rightBlock.label()));

        setCurrentBlock(rightBlock);
        emit(new Compare(result, Compare.Predicate.NE, emitExpr(binary.right()), Constant.of(0)));
        emitBranchIfOpen(endBlock.label());

        setCurrentBlock(endBlock);
        return result;
    }

    private IRValue emitCall(Expr.FuncCall call) {
        FunctionSymbol function = semanticResult.functionOf(call);
        Type returnType = toCommonType(function.returnType());
        List<IRValue> args = new ArrayList<>();
        for (Expr arg : call.args()) {
            args.add(emitExpr(arg));
        }
        Temp result = returnType == Type.VOID ? null : newTemp(returnType);
        emit(new Call(result, function.name(), returnType, args));
        return result;
    }

    private void emitCondExpr(Expr expr, Label trueTarget, Label falseTarget) {
        if (expr instanceof Expr.IntLiteral literal) {
            emit(new Branch(literal.value() != 0 ? trueTarget : falseTarget));
            return;
        }
        if (expr instanceof Expr.Unary unary && unary.op() == Expr.UnaryOp.NOT) {
            emitCondExpr(unary.expr(), falseTarget, trueTarget);
            return;
        }
        if (expr instanceof Expr.Binary binary) {
            if (binary.op() == Expr.BinaryOp.AND) {
                BasicBlock rhsBlock = appendBlock("land.rhs");
                emitCondExpr(binary.left(), rhsBlock.label(), falseTarget);
                setCurrentBlock(rhsBlock);
                emitCondExpr(binary.right(), trueTarget, falseTarget);
                return;
            }
            if (binary.op() == Expr.BinaryOp.OR) {
                BasicBlock rhsBlock = appendBlock("lor.rhs");
                emitCondExpr(binary.left(), trueTarget, rhsBlock.label());
                setCurrentBlock(rhsBlock);
                emitCondExpr(binary.right(), trueTarget, falseTarget);
                return;
            }
        }

        emit(new CondBranch(emitExpr(expr), trueTarget, falseTarget));
    }

    private Temp emitLoadImm(int value) {
        Temp result = newTemp(Type.INT);
        emit(new LoadImm(result, value));
        return result;
    }

    private IRValue addressOf(Symbol symbol) {
        return addressOf(symbolValues.get(symbol));
    }

    private IRValue addressOf(IRValue value) {
        if (value instanceof GlobalVar global) {
            Temp address = newTemp(Type.INT);
            emit(new GlobalAddr(address, global));
            return address;
        }
        return value;
    }

    private int constValueOrZero(Expr expr) {
        OptionalInt value = semanticResult.constValueOf(expr);
        return value.orElse(0);
    }

    private Type toCommonType(FuncDef.Type type) {
        return switch (type) {
            case INT -> Type.INT;
            case VOID -> Type.VOID;
        };
    }

    private record LoopTarget(Label continueTarget, Label breakTarget) {
    }
}

File: src/ir/ir_builder.cpp
#include "toyc/ir/ir_builder.h"
#include <sstream>
#include <stdexcept>
#include <optional>

namespace {
std::optional<int32_t> eval_module_constant(const Expr& expr);

bool eval_binary(int32_t& result, BinaryOp op, int32_t left, int32_t right) {
    auto uleft = static_cast<uint32_t>(left);
    auto uright = static_cast<uint32_t>(right);
    switch (op) {
    case BinaryOp::Add: result = static_cast<int32_t>(uleft + uright); return true;
    case BinaryOp::Sub: result = static_cast<int32_t>(uleft - uright); return true;
    case BinaryOp::Mul: result = static_cast<int32_t>(uleft * uright); return true;
    case BinaryOp::Div:
        if (uright == 0) return false;
        result = static_cast<int32_t>(uleft / uright); return true;
    case BinaryOp::Mod:
        if (uright == 0) return false;
        result = static_cast<int32_t>(uleft % uright); return true;
    case BinaryOp::Lt: result = (left < right) ? 1 : 0; return true;
    case BinaryOp::Gt: result = (left > right) ? 1 : 0; return true;
    case BinaryOp::Le: result = (left <= right) ? 1 : 0; return true;
    case BinaryOp::Ge: result = (left >= right) ? 1 : 0; return true;
    case BinaryOp::Eq: result = (left == right) ? 1 : 0; return true;
    case BinaryOp::Ne: result = (left != right) ? 1 : 0; return true;
    default: return false;
    }
}

std::optional<int32_t> eval_module_constant(const Expr& expr) {
    if (auto* il = dynamic_cast<const IntLiteralExpr*>(&expr)) return il->value;
    if (auto* u = dynamic_cast<const UnaryExpr*>(&expr)) {
        auto operand = eval_module_constant(*u->operand);
        if (!operand) return std::nullopt;
        switch (u->op) {
        case UnaryOp::Plus: return operand;
        case UnaryOp::Minus: return -static_cast<int32_t>(*operand);
        case UnaryOp::Not: return *operand == 0 ? int32_t{1} : int32_t{0};
        }
    }
    if (auto* b = dynamic_cast<const BinaryExpr*>(&expr)) {
        if (b->op == BinaryOp::And) {
            auto lhs = eval_module_constant(*b->left);
            if (!lhs) return std::nullopt;
            if (*lhs == 0) return int32_t{0};
            auto rhs = eval_module_constant(*b->right);
            if (!rhs) return std::nullopt;
            return *rhs != 0 ? int32_t{1} : int32_t{0};
        }
        if (b->op == BinaryOp::Or) {
            auto lhs = eval_module_constant(*b->left);
            if (!lhs) return std::nullopt;
            if (*lhs != 0) return int32_t{1};
            auto rhs = eval_module_constant(*b->right);
            if (!rhs) return std::nullopt;
            return *rhs != 0 ? int32_t{1} : int32_t{0};
        }
        auto lhs = eval_module_constant(*b->left);
        auto rhs = eval_module_constant(*b->right);
        if (!lhs || !rhs) return std::nullopt;
        int32_t result = 0;
        if (eval_binary(result, b->op, *lhs, *rhs)) return result;
    }
    return std::nullopt;
}
}

IRBuilder::IRBuilder() : module_(std::make_unique<Module>()), short_circuit_enabled_(true) {}

std::unique_ptr<IRProgram> IRBuilder::build(const Program& program) {
    IRBuilder builder;
    for (const auto& g : program.globals) {
        int32_t init_val = 0;
        if (g.initializer) {
            auto val = eval_module_constant(*g.initializer);
            if (!val) throw std::runtime_error("global initializer is not a constant integer expression");
            init_val = *val;
        }
        builder.module_->new_global_var(g.name, init_val);
    }
    for (const auto& func : program.functions)
        builder.emit_function(func);
    return std::make_unique<IRProgram>(std::move(builder.module_));
}

void IRBuilder::push_scope() { scopes_.emplace_back(); }
void IRBuilder::pop_scope() { if (!scopes_.empty()) scopes_.pop_back(); }

Value* IRBuilder::lookup_variable(const std::string& name) const {
    for (auto it = scopes_.rbegin(); it != scopes_.rend(); ++it) {
        auto f = it->find(name); if (f != it->end()) return f->second.address;
    }
    return nullptr;
}
void IRBuilder::declare_variable(const std::string& name, Value* addr, const SourceLocation& loc) {
    if (scopes_.empty()) push_scope();
    scopes_.back()[name] = {addr, loc};
}

void IRBuilder::emit_function(const FunctionDef& func) {
    fn_ = new Function(func.name, func.returnType, std::vector<LocalVar*>{});
    module_->add_function(std::unique_ptr<Function>(fn_));
    entry_block_ = fn_->entry_block();
    entry_alloca_index_ = 0;
    current_block_ = entry_block_;

    // Create entry allocas and parameter LocalVars for each function parameter.
    std::vector<LocalVar*> param_locals;
    for (size_t i = 0; i < func.params.size(); i++) {
        LocalVar* addr = fn_->new_local(func.params[i].name, true);
        auto alloca = std::make_unique<AllocaInstr>(addr);
        entry_block_->insert_instruction(entry_alloca_index_++, std::move(alloca));
        param_locals.push_back(addr);
    }
    // Replace the empty params with the actual param address locals
    fn_->set_parameters(param_locals);
    push_scope();
    for (size_t i = 0; i < param_locals.size(); i++)
        declare_variable(func.params[i].name, param_locals[i], SourceLocation{0,0});

    EmitFlow flow = func.body ? emit_compound(*func.body) : EmitFlow::FallsThrough;
    if (flow == EmitFlow::FallsThrough && current_block_ && !current_block_->is_terminated()) {
        Value* def = func.returnType == Type::Void ? nullptr : emit_load_imm(0);
        current_block_->set_terminator(std::make_unique<ReturnInstr>(def));
    }
    pop_scope();
    fn_ = nullptr; entry_block_ = nullptr; current_block_ = nullptr;
}

Value* IRBuilder::emit_expr(const Expr& expr) {
    if (auto* c = dynamic_cast<const CallExpr*>(&expr)) return emit_call(*c);
    if (auto* id = dynamic_cast<const IdentifierExpr*>(&expr)) {
        Value* addr = lookup_variable(id->name);
        if (!addr) {
            // Global variable: emit GlobalAddr + Load
            Value* gaddr = emit_global_addr(id->name);
            return emit_load(gaddr);
        }
        return emit_load(addr);
    }
    if (auto* il = dynamic_cast<const IntLiteralExpr*>(&expr)) return emit_load_imm(il->value);
    if (auto* u = dynamic_cast<const UnaryExpr*>(&expr)) return emit_unary(*u);
    if (auto* b = dynamic_cast<const BinaryExpr*>(&expr)) return emit_binary(*b);
    throw std::runtime_error("unknown expression type");
}

Value* IRBuilder::emit_load_imm(int32_t value) {
    Temp* t = fn_->new_temp(Type::Int);
    current_block_->add_instruction(std::make_unique<LoadImmInstr>(t, fn_->new_constant(value)));
    return t;
}

Value* IRBuilder::emit_unary(const UnaryExpr& expr) {
    Value* op = emit_expr(*expr.operand);
    if (expr.op == UnaryOp::Plus) return op;
    Temp* r = fn_->new_temp(Type::Int);
    current_block_->add_instruction(
        std::make_unique<UnaryOpInstr>(r,
            expr.op == UnaryOp::Minus ? UnaryOpInstr::Op::Neg : UnaryOpInstr::Op::Not, op));
    return r;
}

Value* IRBuilder::emit_binary(const BinaryExpr& expr) {
    if (expr.op == BinaryOp::And || expr.op == BinaryOp::Or) return emit_logical_value(expr);
    bool is_cmp = (expr.op >= BinaryOp::Lt && expr.op <= BinaryOp::Ne);
    Value* l = emit_expr(*expr.left); Value* r = emit_expr(*expr.right);
    Temp* t = fn_->new_temp(Type::Int);
    if (is_cmp) {
        CompareInstr::Predicate p;
        switch (expr.op) {
        case BinaryOp::Lt: p = CompareInstr::Predicate::Lt; break;
        case BinaryOp::Gt: p = CompareInstr::Predicate::Gt; break;
        case BinaryOp::Le: p = CompareInstr::Predicate::Le; break;
        case BinaryOp::Ge: p = CompareInstr::Predicate::Ge; break;
        case BinaryOp::Eq: p = CompareInstr::Predicate::Eq; break;
        case BinaryOp::Ne: p = CompareInstr::Predicate::Ne; break;
        default: return nullptr;
        }
        current_block_->add_instruction(std::make_unique<CompareInstr>(t, p, l, r));
    } else {
        BinaryOpInstr::Op aop;
        switch (expr.op) {
        case BinaryOp::Add: aop = BinaryOpInstr::Op::Add; break;
        case BinaryOp::Sub: aop = BinaryOpInstr::Op::Sub; break;
        case BinaryOp::Mul: aop = BinaryOpInstr::Op::Mul; break;
        case BinaryOp::Div: aop = BinaryOpInstr::Op::Div; break;
        case BinaryOp::Mod: aop = BinaryOpInstr::Op::Mod; break;
        default: throw std::runtime_error("unexpected binary op");
        }
        current_block_->add_instruction(std::make_unique<BinaryOpInstr>(t, aop, l, r));
    }
    return t;
}

Value* IRBuilder::emit_call(const CallExpr& expr) {
    if (expr.return_type == Type::Void) {
        std::vector<Value*> arg_vals;
        for (const auto& arg : expr.args) arg_vals.push_back(emit_expr(*arg));
        current_block_->add_instruction(
            std::make_unique<CallInstr>(nullptr, expr.callee, Type::Void, arg_vals));
        return nullptr;
    }
    std::vector<Value*> arg_vals;
    for (const auto& arg : expr.args) arg_vals.push_back(emit_expr(*arg));
    Temp* result = fn_->new_temp(Type::Int);
    current_block_->add_instruction(std::make_unique<CallInstr>(result, expr.callee, Type::Int, arg_vals));
    return result;
}

// Short-circuit (P1-B)
Value* IRBuilder::emit_logical_value(const BinaryExpr& expr) {
    Value* slot = create_entry_alloca("logic.slot");
    emit_store(emit_load_imm(expr.op == BinaryOp::And ? 0 : 1), slot);
    auto* rhs_b = append_block(expr.op == BinaryOp::And ? "land.rhs" : "lor.rhs");
    auto* merge_b = append_block(expr.op == BinaryOp::And ? "land.end" : "lor.end");
    if (expr.op == BinaryOp::And) emit_condition(*expr.left, rhs_b->label(), merge_b->label());
    else emit_condition(*expr.left, merge_b->label(), rhs_b->label());
    current_block_ = rhs_b;
    Value* rv = emit_expr(*expr.right); emit_store(normalize_bool(rv), slot);
    if (!current_block_->is_terminated())
        current_block_->set_terminator(std::make_unique<BranchInstr>(merge_b->label()));
    current_block_ = merge_b; return emit_load(slot);
}
void IRBuilder::emit_condition(const Expr& expr, Label* tt, Label* ft) {
    if (auto* b = dynamic_cast<const BinaryExpr*>(&expr)) {
        if (b->op == BinaryOp::And) { auto* r = append_block("land.rhs");
            emit_condition(*b->left, r->label(), ft); current_block_ = r;
            emit_condition(*b->right, tt, ft); return; }
        if (b->op == BinaryOp::Or) { auto* r = append_block("lor.rhs");
            emit_condition(*b->left, tt, r->label()); current_block_ = r;
            emit_condition(*b->right, tt, ft); return; }
    }
    if (auto* u = dynamic_cast<const UnaryExpr*>(&expr)) {
        if (u->op == UnaryOp::Not) { emit_condition(*u->operand, ft, tt); return; }
    }
    if (auto* il = dynamic_cast<const IntLiteralExpr*>(&expr)) {
        current_block_->set_terminator(std::make_unique<BranchInstr>(il->value ? tt : ft)); return;
    }
    Value* v = emit_expr(expr); Temp* c = fn_->new_temp(Type::Int);
    current_block_->add_instruction(std::make_unique<CompareInstr>(c, CompareInstr::Predicate::Ne, v, fn_->new_constant(0)));
    current_block_->set_terminator(std::make_unique<CondBranchInstr>(c, tt, ft));
}
Value* IRBuilder::normalize_bool(Value* val) {
    Temp* r = fn_->new_temp(Type::Int);
    current_block_->add_instruction(std::make_unique<CompareInstr>(r, CompareInstr::Predicate::Ne, val, fn_->new_constant(0)));
    return r;
}

// Statements (P2-A)
EmitFlow IRBuilder::emit_stmt(const Stmt& stmt) {
    if (auto* c = dynamic_cast<const CompoundStmt*>(&stmt)) return emit_compound(*c);
    if (auto* d = dynamic_cast<const VarDeclStmt*>(&stmt)) return emit_var_decl(*d);
    if (auto* a = dynamic_cast<const AssignStmt*>(&stmt)) return emit_assign(*a);
    if (auto* i = dynamic_cast<const IfStmt*>(&stmt)) return emit_if(*i);
    if (auto* w = dynamic_cast<const WhileStmt*>(&stmt)) return emit_while(*w);
    if (dynamic_cast<const BreakStmt*>(&stmt)) return emit_break();
    if (dynamic_cast<const ContinueStmt*>(&stmt)) return emit_continue();
    if (auto* r = dynamic_cast<const ReturnStmt*>(&stmt)) return emit_return(*r);
    if (auto* e = dynamic_cast<const ExprStmt*>(&stmt)) return emit_expr_stmt(*e);
    return EmitFlow::FallsThrough;
}
EmitFlow IRBuilder::emit_compound(const CompoundStmt& stmt) {
    push_scope(); EmitFlow flow = EmitFlow::FallsThrough;
    for (const auto& s : stmt.statements) {
        if (flow == EmitFlow::Terminates) continue;
        flow = emit_stmt(*s);
    } pop_scope(); return flow;
}
EmitFlow IRBuilder::emit_var_decl(const VarDeclStmt& stmt) {
    Value* addr = create_entry_alloca(stmt.name);
    declare_variable(stmt.name, addr, stmt.location);
    if (stmt.initializer && current_block_ && !current_block_->is_terminated())
        emit_store(emit_expr(*stmt.initializer), addr);
    return EmitFlow::FallsThrough;
}
EmitFlow IRBuilder::emit_assign(const AssignStmt& stmt) {
    Value* addr = lookup_variable(stmt.name);
    if (!addr) {
        // Global assignment
        addr = emit_global_addr(stmt.name);
    }
    if (current_block_ && !current_block_->is_terminated()) emit_store(emit_expr(*stmt.value), addr);
    return EmitFlow::FallsThrough;
}
EmitFlow IRBuilder::emit_if(const IfStmt& stmt) {
    auto* then_b = append_block("if.then"); BasicBlock* else_b = nullptr; BasicBlock* merge_b = nullptr;
    if (stmt.elseStmt) { else_b = append_block("if.else");
        emit_condition(*stmt.condition, then_b->label(), else_b->label()); }
    else { merge_b = append_block("if.end");
        emit_condition(*stmt.condition, then_b->label(), merge_b->label()); }
    current_block_ = then_b; EmitFlow then_f = emit_stmt(*stmt.thenStmt);
    if (then_f == EmitFlow::FallsThrough && !current_block_->is_terminated()) {
        if (!merge_b) merge_b = append_block("if.end");
        current_block_->set_terminator(std::make_unique<BranchInstr>(merge_b->label()));
    }
    EmitFlow else_f = EmitFlow::Terminates;
    if (stmt.elseStmt) { current_block_ = else_b; else_f = emit_stmt(*stmt.elseStmt);
        if (else_f == EmitFlow::FallsThrough && !current_block_->is_terminated()) {
            if (!merge_b) merge_b = append_block("if.end");
            current_block_->set_terminator(std::make_unique<BranchInstr>(merge_b->label())); }
        }
    else else_f = EmitFlow::FallsThrough;
    if (then_f == EmitFlow::Terminates && else_f == EmitFlow::Terminates) {
        current_block_ = nullptr;
        return EmitFlow::Terminates;
    }
    current_block_ = merge_b; return EmitFlow::FallsThrough;
}
EmitFlow IRBuilder::emit_while(const WhileStmt& stmt) {
    auto* cond_b = append_block("while.cond"); auto* body_b = append_block("while.body"); auto* exit_b = append_block("while.end");
    if (current_block_ && !current_block_->is_terminated())
        current_block_->set_terminator(std::make_unique<BranchInstr>(cond_b->label()));
    current_block_ = cond_b; emit_condition(*stmt.condition, body_b->label(), exit_b->label());
    current_block_ = body_b; loop_stack_.push_back({exit_b, cond_b});
    EmitFlow body_f = emit_stmt(*stmt.body); loop_stack_.pop_back();
    if (body_f == EmitFlow::FallsThrough && current_block_ && !current_block_->is_terminated())
        current_block_->set_terminator(std::make_unique<BranchInstr>(cond_b->label()));
    current_block_ = exit_b; return EmitFlow::FallsThrough;
}
EmitFlow IRBuilder::emit_break() {
    if (!loop_stack_.empty() && current_block_ && !current_block_->is_terminated())
        current_block_->set_terminator(std::make_unique<BranchInstr>(loop_stack_.back().break_target->label()));
    return EmitFlow::Terminates;
}
EmitFlow IRBuilder::emit_continue() {
    if (!loop_stack_.empty() && current_block_ && !current_block_->is_terminated())
        current_block_->set_terminator(std::make_unique<BranchInstr>(loop_stack_.back().continue_target->label()));
    return EmitFlow::Terminates;
}
EmitFlow IRBuilder::emit_return(const ReturnStmt& stmt) {
    Value* val = stmt.value ? emit_expr(*stmt.value) : nullptr;
    if (current_block_ && !current_block_->is_terminated())
        current_block_->set_terminator(std::make_unique<ReturnInstr>(val));
    return EmitFlow::Terminates;
}
EmitFlow IRBuilder::emit_expr_stmt(const ExprStmt& stmt) { emit_expr(*stmt.expr); return EmitFlow::FallsThrough; }

Value* IRBuilder::create_entry_alloca(const std::string& hint) {
    LocalVar* l = fn_->new_local(hint, false);
    auto a = std::make_unique<AllocaInstr>(l); Value* addr = a->result();
    entry_block_->insert_instruction(entry_alloca_index_++, std::move(a));
    return addr;
}
Value* IRBuilder::emit_load(Value* addr) {
    Temp* r = fn_->new_temp(Type::Int);
    current_block_->add_instruction(std::make_unique<LoadInstr>(r, addr));
    return r;
}
void IRBuilder::emit_store(Value* val, Value* addr) {
    current_block_->add_instruction(std::make_unique<StoreInstr>(val, addr));
}
BasicBlock* IRBuilder::append_block(const std::string& hint) {
    auto* l = fn_->new_label(hint); auto b = std::make_unique<BasicBlock>(l);
    auto* raw = b.get(); fn_->add_block(std::move(b)); return raw;
}
Value* IRBuilder::emit_global_addr(const std::string& name) {
    // Find the GlobalVar in the module
    GlobalVar* gv = nullptr;
    for (auto* g : module_->globals()) {
        if (g->symbol_name() == name) { gv = g; break; }
    }
    if (!gv) throw std::runtime_error("global variable '" + name + "' not found");
    Temp* addr = fn_->new_temp(Type::Int);
    current_block_->add_instruction(std::make_unique<GlobalAddrInstr>(addr, gv));
    return addr;
}

void IRBuilder::emit(std::unique_ptr<Instr> instr) { current_block_->add_instruction(std::move(instr)); }
